lib/puppet/resource/catalog.rb

```diff
@@ -366,18 +369,14 @@ def resource(type, title = nil)
       # an instance has to be created in order to construct the unique key used when
       # searching for aliases, or when app_management is active and nothing is found in
       # which case it is needed by the CapabilityFinder.
-# PUP-4947
       res = nil
-#      app_mgnt = Puppet[:app_management]
-#      if app_mgnt || !@aliases.empty?
       res = Puppet::Resource.new(type, title, { :environment => @environment_instance })
 
-        # No need to build the uniqueness key unless there are aliases
-      result = @resource_table[[type_name, res.uniqueness_key].flatten] unless @aliases.empty?
-# PUP-4947
-#      end
+      # Must check with uniqueness key because of aliases or if resource transforms title in title
+      # to attribute mappings.
+      result = @resource_table[[type_name, res.uniqueness_key].flatten]
```

Comment:
Oh, I get it now. It seems less desirable to compile the catalog only to fail later on the agent. In the future, should we fail compile too? If so, can we put a deprecation warning when strict mode is enabled? I'm not sure if we have enough information at compile time to detect the problem given it is `File` specific behavior?